pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -85,48 +90,69 @@ func (s *SSTSinkKeyWriter) WriteKey(ctx context.Context, key storage.MVCCKey, va
 		return err
 	}
 
-	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedFiles[len(s.flushedFiles)-1].EntryCounts.Add(keyAsRowCount)
 	s.flushedSize += keyAsRowCount.DataSize
 	s.prevKey = append(s.prevKey[:0], key.Key...)
-
+	// Until the next key is written, we cannot determine if this key is mid-row.
+	// As a safeguard, we assume it is mid-row, and if it is the last key to be
+	// written in this span, the caller has the responsibility of explicitly
+	// marking the key as not mid-row by calling AssumeNotMidRow.
+	s.midRow = true
 	return nil
 }
 
+// AssumeNotMidRow marks the last key written as not mid-row. This exists to
+// ensure that the caller explicitly is aware that SSTSinkKeyWriter is unable
+// to determine if the last key written in a span is mid-row and must enforce
+// the invariant themselves.
+func (s *SSTSinkKeyWriter) AssumeNotMidRow() {
+	s.midRow = false
+}
+
 // Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
 // open BackupManifest_File exists for the new span. It will either open a new
 // BackupManifest_File, or if the new span contiguously extends the last span,
 // it will extend it if it remains under the fileSpanByteLimit. The caller is
-// responsible for ensuring that spans that are reset on do not end mid-row.
+// responsible for ensuring that spans that are reset on do not end mid-row
+// and calling AssumeNotMidRow before resetting to enforce this invariant.
 // Any time a new span is being written, Reset MUST be called prior to any
 // WriteKey calls.
 func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
 	log.VEventf(ctx, 2, "resetting sink to span %s", newSpan)
+	if s.midRow {
+		return errors.AssertionFailedf("cannot reset after writing a mid-row key")
+	}
 	if err := s.maybeFlushNonExtendableSpans(newSpan); err != nil {
```

Comment:
nit: we should rename this to `maybeFlushNonExtendibleSST` as this check is really about asserting backup SST invariants: keys must be written in order and keys must have the same prefix. 

Further, this current name implies that we're extending the last span entry in the SST, but that is not what it does.